add fallthrough kernel for AutogradMeta key

bdhirsh opened the pull request:
The other `Autograd[Backend]` keys all have fallthrough kernels registered to them, but `AutogradMeta` was missing the fallthrough kernel.

This is a problem for custom ops that don't have autograd support, if you try to run them with meta tensors. If you have a custom op, and register a CPU and a Meta kernel, then:

(1) if you run the op with cpu tensors, it will dispatch straight to the CPU kernel (as expected)

(2) if you run the op with meta tensors, you will error - because we don't have a fallthrough registered to the AutogradMeta key, we will try to dispatch to the AutogradMeta key and error, since the op author hasn't provided an autograd implementation.

Here's a repro that I confirmed now works:

```
import torch
from torch._dispatch.python import enable_python_dispatcher
from torch._subclasses.fake_tensor import FakeTensorMode

lib = torch.library.Library("test", "DEF")
impl_cpu = torch.library.Library("test", "IMPL", "CPU")
impl_meta = torch.library.Library("test", "IMPL", "Meta")

def foo_impl(x):
    return x + 1

lib.define("foo(Tensor a) -> Tensor")
impl_meta.impl("foo", foo_impl)
impl_cpu.impl("foo", foo_impl)


with enable_python_dispatcher():
    a = torch.ones(2, device='meta')
    print("@@@@@")
    b = torch.ops.test.foo.default(a)
    print(b)
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94603

bdhirsh committed:
add fallthrough kernel for AutogradMeta key

[ghstack-poisoned]

ezyang reviewed (approved):
A test would be good. Also, if someone adds a new device key, they'll run into the same problem; any way to easily (**) make this future proof?

bdhirsh commented:
> A test would be good. Also, if someone adds a new device key, they'll run into the same problem; any way to easily (**) make this future proof?

I can put a version of my repro into a test somewhere.

What's annoying is that we don't want to register a fallthrough to **all** autograd keys (the comment in that file explicitly says that PrivateUse keys shouldn't get a fallthrough).

We could probably macro-ify it. Although going forward, I think we should prefer that new backends integrate through the existing PrivateUse1 key instead of adding new backend keys. Either way, I can add a comment in the BackendComponent enum that tells new backends about the situation.

albanD reviewed (approved):
Should this file just be codegened to make it more future proof?

bdhirsh commented:
We could, although codegen feels a bit like overkill to me (new devices should be added very rarely, and now we have a new codegen'd C++ file that we need to maintain + make sure gets picked up properly in our internal build systems)

bdhirsh committed:
Update on "add fallthrough kernel for AutogradMeta key"

The other `Autograd[Backend]` keys all have fallthrough kernels registered to them, but `AutogradMeta` was missing the fallthrough kernel.

This is a problem for custom ops that don't have autograd support, if you try to run them with meta tensors. If you have a custom op, and register a CPU and a Meta kernel, then:

(1) if you run the op with cpu tensors, it will dispatch straight to the CPU kernel (as expected)

(2) if you run the op with meta tensors, you will error - because we don't have a fallthrough registered to the AutogradMeta key, we will try to dispatch to the AutogradMeta key and error, since the op author hasn't provided an autograd implementation.

Here's a repro that I confirmed now works:

```
import torch
from torch._dispatch.python import enable_python_dispatcher
from torch._subclasses.fake_tensor import FakeTensorMode

lib = torch.library.Library("test", "DEF")
impl_cpu = torch.library.Library("test", "IMPL", "CPU")
impl_meta = torch.library.Library("test", "IMPL", "Meta")

def foo_impl(x):
    return x + 1

lib.define("foo(Tensor a) -> Tensor")
impl_meta.impl("foo", foo_impl)
impl_cpu.impl("foo", foo_impl)


with enable_python_dispatcher():
    a = torch.ones(2, device='meta')
    print("@@@@@")
    b = torch.ops.test.foo.default(a)
    print(b)
```




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge -f "trunk failures"
